### ISSUE TITLE
[🐛BUG] Strange learning of context-aware FM and DeepFM on Yelp dataset

### ISSUE THREAD

**Issue by adebiasio-estilos** [RUCAIBox/RecBole#1552]
I am training **top-k** recommendation models on a custom **Yelp** dataset organized as follows:

**Yelp_recsys.inter**

```
	user_id:token			item_id:token			rating:float
0	mh_-eMZ6K5RLWhZyISBhwA		XQfwVwDr-v0ZS3_CbbE5Xw		3
1	Iaee7y6zdSB3B-kRCo4z1w		XQfwVwDr-v0ZS3_CbbE5Xw		2
2	ejFxLGqQcWNLdNByJlIhnQ		XQfwVwDr-v0ZS3_CbbE5Xw		4
3	f7xa0p_1V9lx53iIGN5Sug		XQfwVwDr-v0ZS3_CbbE5Xw		3
4	dCooFVCk8M1nVaQqcfTL3Q		XQfwVwDr-v0ZS3_CbbE5Xw		2
```

**Yelp_recsys.item**

```
	item_id:token			category:token_seq	subcategory:token_seq	price_bucket:token_seq
0	--epgcb7xHGuJ-4PUeSLAw		Restaurants		Food			1.0
1	-0EdehHjIQc0DtYU8QcAig		Restaurants		Chinese			NaN
2	-0G_6-KFGpCpxTUlVXCMYQ		Restaurants		Mexican			2.0
3	-0PN_KFPtbnLQZEeb23XiA		Restaurants		Chinese			2.0
4	-0dKgi_Hpcis921nOpM85Q		Restaurants		Barbeque		2.0
```

With models belonging to the **general_recommendations** set, there are no problems and learning seems ok (also using NGCF with item features).

With models belonging to the **context_aware** set (i.e., FM, DeepFM) it appears that the algorithm fails to learn even if the loss is descending (using 200 epochs results don't change).

How should I change the configurations used for FM/DeepFM?

Follows the configuration and results using BPR anf FM.
---------------------------------------------------------------------------------------------------------

**FM**
```
General Hyper Parameters:
gpu_id = 0
use_gpu = True
seed = 123
state = WARNING
reproducibility = True
data_path = /tmp/tmpznbnvdms/yelp_recsys
checkpoint_dir = saved/
show_progress = True
save_dataset = False
dataset_save_path = None
save_dataloaders = False
dataloaders_save_path = None
log_wandb = False

Training Hyper Parameters:
epochs = 10
train_batch_size = 1024
learner = adam
learning_rate = 0.001
neg_sampling = None
eval_step = 5
stopping_step = 10
clip_grad_norm = None
weight_decay = 0.0
loss_decimal_place = 4

Evaluation Hyper Parameters:
eval_args = {'group_by': 'user', 'order': 'RO', 'split': {'RS': [0.6, 0.2, 0.2]}, 'mode': 'full'}
repeatable = False
metrics = ['Recall', 'NDCG']
topk = [10, 50, 100]
valid_metric = NDCG@100
valid_metric_bigger = True
eval_batch_size = 1000000
metric_decimal_place = 4

Dataset Hyper Parameters:
field_separator = 	
seq_separator =  
USER_ID_FIELD = user_id
ITEM_ID_FIELD = item_id
RATING_FIELD = rating
TIME_FIELD = timestamp
seq_len = None
LABEL_FIELD = label
threshold = {'rating': 4}
NEG_PREFIX = neg_
load_col = {'inter': ['user_id', 'item_id', 'rating'], 'item': ['item_id', 'price_bucket', 'category', 'subcategory']}
unload_col = None
unused_col = None
additional_feat_suffix = None
rm_dup_inter = None
val_interval = None
filter_inter_by_user_or_item = True
user_inter_num_interval = [0,inf)
item_inter_num_interval = [0,inf)
alias_of_user_id = None
alias_of_item_id = None
alias_of_entity_id = None
alias_of_relation_id = None
preload_weight = None
normalize_field = None
normalize_all = None
ITEM_LIST_LENGTH_FIELD = item_length
LIST_SUFFIX = _list
MAX_ITEM_LIST_LENGTH = 50
POSITION_FIELD = position_id
HEAD_ENTITY_ID_FIELD = head_id
TAIL_ENTITY_ID_FIELD = tail_id
RELATION_ID_FIELD = relation_id
ENTITY_ID_FIELD = entity_id
benchmark_filename = None

Other Hyper Parameters: 
wandb_project = recbole
require_pow = False
embedding_size = 64
MODEL_TYPE = ModelType.CONTEXT
worker = 12
shuffle = True
MODEL_INPUT_TYPE = InputType.POINTWISE
eval_type = EvaluatorType.RANKING
device = cuda
train_neg_sample_args = {'strategy': 'none'}
eval_neg_sample_args = {'strategy': 'full', 'distribution': 'uniform'}

21 Nov 15:31    INFO  yelp_recsys
The number of users: 589050
Average actions of users: 2.021210459571275
The number of items: 15177
Average actions of items: 78.45229309435952
The number of inters: 1190592
The sparsity of the dataset: 99.98668243375987%
Remain Fields: ['user_id', 'item_id', 'category', 'subcategory', 'price_bucket', 'label']
21 Nov 15:31    INFO  [Training]: train_batch_size = [1024] negative sampling: [None]
21 Nov 15:31    INFO  [Evaluation]: eval_batch_size = [1000000] eval_args: [{'group_by': 'user', 'order': 'RO', 'split': {'RS': [0.6, 0.2, 0.2]}, 'mode': 'full'}]
21 Nov 15:31    INFO  FM(
  (token_embedding_table): FMEmbedding(
    (embedding): Embedding(604227, 64)
  )
  (token_seq_embedding_table): ModuleList(
    (0): Embedding(2, 64)
    (1): Embedding(306, 64)
    (2): Embedding(5, 64)
  )
  (first_order_linear): FMFirstOrderLinear(
    (token_embedding_table): FMEmbedding(
      (embedding): Embedding(604227, 1)
    )
    (token_seq_embedding_table): ModuleList(
      (0): Embedding(2, 1)
      (1): Embedding(306, 1)
      (2): Embedding(5, 1)
    )
  )
  (fm): BaseFactorizationMachine()
  (sigmoid): Sigmoid()
  (loss): BCELoss()
)
Trainable parameters: 39295101

21 Nov 15:31    INFO  epoch 0 training [time: 8.73s, train loss: 485.4505]
21 Nov 15:31    INFO  epoch 1 training [time: 8.82s, train loss: 351.5899]
21 Nov 15:31    INFO  epoch 2 training [time: 8.65s, train loss: 202.3874]
21 Nov 15:31    INFO  epoch 3 training [time: 9.00s, train loss: 146.0716]
21 Nov 15:32    INFO  epoch 4 training [time: 8.88s, train loss: 125.0351]
21 Nov 15:33    INFO  epoch 4 evaluating [time: 85.33s, valid_score: 0.000500]
21 Nov 15:33    INFO  valid result: 
recall@10 : 0.0004    recall@50 : 0.0012    recall@100 : 0.0019    ndcg@10 : 0.0002    ndcg@50 : 0.0004    ndcg@100 : 0.0005
21 Nov 15:33    INFO  epoch 5 training [time: 8.58s, train loss: 111.3253]
21 Nov 15:33    INFO  epoch 6 training [time: 8.90s, train loss: 100.3269]
21 Nov 15:33    INFO  epoch 7 training [time: 8.58s, train loss: 90.8923]
21 Nov 15:34    INFO  epoch 8 training [time: 8.55s, train loss: 82.7699]
21 Nov 15:34    INFO  epoch 9 training [time: 8.59s, train loss: 75.6053]
21 Nov 15:35    INFO  epoch 9 evaluating [time: 85.03s, valid_score: 0.000500]
21 Nov 15:35    INFO  valid result: 
recall@10 : 0.0003    recall@50 : 0.0009    recall@100 : 0.0019    ndcg@10 : 0.0001    ndcg@50 : 0.0003    ndcg@100 : 0.0005

```

**BPR**
```
General Hyper Parameters:
gpu_id = 0
use_gpu = True
seed = 123
state = INFO
reproducibility = True
data_path = /tmp/tmpznbnvdms/yelp_recsys
checkpoint_dir = saved/
show_progress = True
save_dataset = False
dataset_save_path = None
save_dataloaders = False
dataloaders_save_path = None
log_wandb = False

Training Hyper Parameters:
epochs = 10
train_batch_size = 1024
learner = adam
learning_rate = 0.01
neg_sampling = {'uniform': 1}
eval_step = 5
stopping_step = 10
clip_grad_norm = None
weight_decay = 0.0
loss_decimal_place = 4

Evaluation Hyper Parameters:
eval_args = {'group_by': 'user', 'order': 'RO', 'split': {'RS': [0.6, 0.2, 0.2]}, 'mode': 'full'}
repeatable = False
metrics = ['Recall', 'NDCG']
topk = [10, 50, 100]
valid_metric = NDCG@100
valid_metric_bigger = True
eval_batch_size = 10000000
metric_decimal_place = 4

Dataset Hyper Parameters:
field_separator = 	
seq_separator =  
USER_ID_FIELD = user_id
ITEM_ID_FIELD = item_id
RATING_FIELD = rating
TIME_FIELD = timestamp
seq_len = None
LABEL_FIELD = label
threshold = None
NEG_PREFIX = neg_
load_col = {'inter': ['user_id', 'item_id', 'rating']}
unload_col = None
unused_col = None
additional_feat_suffix = None
rm_dup_inter = None
val_interval = None
filter_inter_by_user_or_item = True
user_inter_num_interval = [0,inf)
item_inter_num_interval = [0,inf)
alias_of_user_id = None
alias_of_item_id = None
alias_of_entity_id = None
alias_of_relation_id = None
preload_weight = None
normalize_field = None
normalize_all = None
ITEM_LIST_LENGTH_FIELD = item_length
LIST_SUFFIX = _list
MAX_ITEM_LIST_LENGTH = 50
POSITION_FIELD = position_id
HEAD_ENTITY_ID_FIELD = head_id
TAIL_ENTITY_ID_FIELD = tail_id
RELATION_ID_FIELD = relation_id
ENTITY_ID_FIELD = entity_id
benchmark_filename = None

Other Hyper Parameters: 
wandb_project = recbole
require_pow = False
embedding_size = 64
MODEL_TYPE = ModelType.GENERAL
worker = 12
shuffle = True
MODEL_INPUT_TYPE = InputType.PAIRWISE
eval_type = EvaluatorType.RANKING
device = cuda
train_neg_sample_args = {'strategy': 'by', 'by': 1, 'distribution': 'uniform', 'dynamic': 'none'}
eval_neg_sample_args = {'strategy': 'full', 'distribution': 'uniform'}

The number of users: 589380
Average actions of users: 2.0221181955923098
The number of items: 15291
Average actions of items: 77.94597776324395
The number of inters: 1191794
The sparsity of the dataset: 99.98677578467944%
Remain Fields: ['user_id', 'item_id', 'rating']
21 Nov 15:27    INFO  [Training]: train_batch_size = [1024] negative sampling: [{'uniform': 1}]
21 Nov 15:27    INFO  [Evaluation]: eval_batch_size = [10000000] eval_args: [{'group_by': 'user', 'order': 'RO', 'split': {'RS': [0.6, 0.2, 0.2]}, 'mode': 'full'}]
21 Nov 15:27    INFO  BPR(
  (user_embedding): Embedding(589380, 64)
  (item_embedding): Embedding(15291, 64)
  (loss): BPRLoss()
)
Trainable parameters: 38698944

21 Nov 15:27    INFO  epoch 0 training [time: 6.59s, train loss: 529.7719]
21 Nov 15:27    INFO  epoch 1 training [time: 6.49s, train loss: 98.6137]
21 Nov 15:28    INFO  epoch 2 training [time: 6.43s, train loss: 12.1346]
21 Nov 15:28    INFO  epoch 3 training [time: 6.44s, train loss: 8.4156]
21 Nov 15:28    INFO  epoch 4 training [time: 6.43s, train loss: 7.1541]
21 Nov 15:28    INFO  epoch 4 evaluating [time: 2.16s, valid_score: 0.097700]
21 Nov 15:28    INFO  valid result: 
recall@10 : 0.088    recall@50 : 0.1947    recall@100 : 0.2739    ndcg@10 : 0.059    ndcg@50 : 0.0837    ndcg@100 : 0.0977
21 Nov 15:28    INFO  epoch 5 training [time: 6.43s, train loss: 6.8639]
21 Nov 15:28    INFO  epoch 6 training [time: 6.44s, train loss: 6.6109]
21 Nov 15:28    INFO  epoch 7 training [time: 6.54s, train loss: 6.7025]
21 Nov 15:28    INFO  epoch 8 training [time: 6.52s, train loss: 6.1265]
21 Nov 15:28    INFO  epoch 9 training [time: 6.83s, train loss: 6.3258]
21 Nov 15:28    INFO  epoch 9 evaluating [time: 2.31s, valid_score: 0.098400]
21 Nov 15:28    INFO  valid result: 
recall@10 : 0.0874    recall@50 : 0.1942    recall@100 : 0.2723    ndcg@10 : 0.06    ndcg@50 : 0.0847    ndcg@100 : 0.0984
```

**Comment by adebiasio-estilos** [comment]
Hi Zheng thanks for answering!
1. Yes, I know that FM can exploit both user and item features but I think that with just item features and not user features it should work as well (e.g., here is used as baseline in this scenario -> Zheng, Yu, et al. "Incorporating price into recommendation with graph convolutional networks." IEEE Transactions on Knowledge and Data Engineering (2021).).
2. Ok, modified it!
3. Yes, I know that FM is traditionally used for rating predictions and evaluated via AUC but in many cases, rating predictions models can be exploited also in ranking tasks (e.g., here is used for ranking in this scenario -> Zheng, Yu, et al. "Incorporating price into recommendation with graph convolutional networks." IEEE Transactions on Knowledge and Data Engineering (2021).).

Do you have other suggestions to use FM in ranking with just item features?

**Comment by zhengbw0324** [comment]
@adebiasio-estilos 
I found that you have not set the negative sampling strategy in the training process. I suggest you set the negative sampling strategy. For example:
```
train_neg_sample_args:
  distribution: uniform
  sample_num: 1
  dynamic: False
```

**Comment by adebiasio-estilos** [comment]
Hi Zheng thanks for your help!
I added as suggested but unfortunately it does not impact the previous outcomes from the model.

I found that similar issues emerged recently https://github.com/RUCAIBox/RecBole/issues/1531 and last year https://github.com/RUCAIBox/RecBole/issues/819 https://github.com/RUCAIBox/RecBole/issues/827 but don't know if actually they have been already addressed or if some custom implementation is needed.

**Comment by adebiasio-estilos** [comment]
I found something interesting!
I think I should set both these configurations to enable negative_sampling:
```
config_settings['neg_sampling'] = {'uniform': 1}
config_settings['train_neg_sample_args'] = {'distribution': 'uniform', 'sample_num': 1, 'dynamic': False} 
```
I should do some other tests but the model seems learning!
